src/phoenix/server/jwt_store.py

```diff
@@ -0,0 +1,397 @@
+import logging
+from abc import ABC, abstractmethod
+from asyncio import create_task, gather, sleep
+from copy import deepcopy
+from dataclasses import replace
+from datetime import datetime, timezone
+from functools import cached_property
+from typing import Any, Dict, Generic, List, Optional, Tuple, Type, TypeVar
+
+import jwt
+from sqlalchemy import Select, delete, insert, select
+
+from phoenix.auth import (
+    JWT_ALGORITHM,
+    ClaimSet,
+    Token,
+)
+from phoenix.db import models
+from phoenix.db.enums import UserRole
+from phoenix.server.types import (
+    AccessToken,
+    AccessTokenAttributes,
+    AccessTokenClaims,
+    AccessTokenId,
+    ApiKey,
+    ApiKeyAttributes,
+    ApiKeyClaims,
+    ApiKeyId,
+    DaemonTask,
+    DbSessionFactory,
+    RefreshToken,
+    RefreshTokenAttributes,
+    RefreshTokenClaims,
+    RefreshTokenId,
+    TokenId,
+    UserId,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class JwtStore:
+    def __init__(
+        self,
+        db: DbSessionFactory,
+        secret: str,
+        algorithm: str = JWT_ALGORITHM,
+        sleep_seconds: int = 10,
+        **kwargs: Any,
+    ) -> None:
+        assert secret
+        super().__init__(**kwargs)
+        self._secret = secret
+        args = (db, secret, algorithm, sleep_seconds)
+        self._access_token_store = AccessTokenStore(*args, **kwargs)
+        self._refresh_token_store = RefreshTokenStore(*args, **kwargs)
+        self._api_key_store = ApiKeyStore(*args, **kwargs)
+
+    @cached_property
+    def _stores(self) -> Tuple[DaemonTask, ...]:
+        return tuple(dt for dt in self.__dict__.values() if isinstance(dt, _Store))
+
+    async def __aenter__(self) -> None:
+        await gather(*(s.__aenter__() for s in self._stores))
+
+    async def __aexit__(self, *args: Any, **kwargs: Any) -> None:
+        await gather(*(s.__aexit__(*args, **kwargs) for s in self._stores))
+
+    async def read(self, token: Token) -> Optional[ClaimSet]:
+        try:
+            payload = jwt.decode(
+                token,
+                self._secret,
+                algorithms=[JWT_ALGORITHM],
+                options={"verify_exp": False},
+            )
+        except jwt.DecodeError:
+            return None
+        if (jti := payload.get("jti")) is None:
+            return None
+        if (token_id := TokenId.parse(jti)) is None:
+            return None
+        if isinstance(token_id, AccessTokenId):
+            return await self._access_token_store.get(token_id)
+        if isinstance(token_id, RefreshTokenId):
+            return await self._refresh_token_store.get(token_id)
+        if isinstance(token_id, ApiKeyId):
+            return await self._api_key_store.get(token_id)
```

Comment:
No, it's an standalone abstraction specific to the token store. it's how how the token are identified. it doesn't have anything to do with graphql

Reply:
What do the IDs actually look like?

I was thinking we would just have a "type" claim in the JWT. Is this a common pattern to encode the type in the ID?

Reply:
they look like this `ApiKey:1`, `AccessToken:1`, etc. where the number is the database primary key

Reply:
> Is this a common pattern

don't know, but it seems to work.

> we would just have a "type" claim

since `jti` should still be unique, it's kind of redundant

Reply:
That makes sense to me. These actually are just the unencoded node IDs, I think, which I like.